bigframes/core/__init__.py

```diff
@@ -49,470 +40,196 @@
-    Ibis projection operations to keep generated SQL small and correct when
-    mixing and matching columns from different versions of a DataFrame.
+    node: nodes.BigFrameNode
 
-    Args:
-        session:
-            A BigQuery DataFrames session to allow more flexibility in running
-            queries.
-        table: An Ibis table expression.
-        columns: Ibis value expressions that can be projected as columns.
-        hidden_ordering_columns: Ibis value expressions to store ordering.
-        ordering: An ordering property of the data frame.
-        predicates: A list of filters on the data frame.
-    """
-
-    def __init__(
-        self,
+    @classmethod
+    def from_ibis(
+        cls,
         session: Session,
         table: ibis_types.Table,
         columns: Sequence[ibis_types.Value],
-        hidden_ordering_columns: Optional[Sequence[ibis_types.Value]] = None,
-        ordering: ExpressionOrdering = ExpressionOrdering(),
-        predicates: Optional[Collection[ibis_types.BooleanValue]] = None,
+        hidden_ordering_columns: Sequence[ibis_types.Value],
+        ordering: orderings.ExpressionOrdering,
     ):
-        self._session = session
-        self._table = table
-        self._predicates = tuple(predicates) if predicates is not None else ()
-        # TODO: Validate ordering
-        if not ordering.total_ordering_columns:
-            raise ValueError("Must have total ordering defined by one or more columns")
-        self._ordering = ordering
-        # Allow creating a DataFrame directly from an Ibis table expression.
-        # TODO(swast): Validate that each column references the same table (or
-        # no table for literal values).
-        self._columns = tuple(columns)
-
-        # Meta columns store ordering, or other data that doesn't correspond to dataframe columns
-        self._hidden_ordering_columns = (
-            tuple(hidden_ordering_columns)
-            if hidden_ordering_columns is not None
-            else ()
+        node = nodes.ReadGbqNode(
+            table=table,
+            session=session,
+            columns=tuple(columns),
+            hidden_ordering_columns=tuple(hidden_ordering_columns),
+            ordering=ordering,
         )
+        return cls(node)
 
-        # To allow for more efficient lookup by column name, create a
-        # dictionary mapping names to column values.
-        self._column_names = {column.get_name(): column for column in self._columns}
-        self._hidden_ordering_column_names = {
-            column.get_name(): column for column in self._hidden_ordering_columns
-        }
-        ### Validation
-        value_col_ids = self._column_names.keys()
-        hidden_col_ids = self._hidden_ordering_column_names.keys()
+    @classmethod
+    def from_pandas(cls, pd_df: pandas.DataFrame):
+        # Need to make immutable and hashable
+        local_array = tuple(tuple(val) for _, val in pd_df.items())  # column-major
```

Comment:
This is going to be quite slow. To be consistent with how we do load jobs, let's create a Parquet file (in memory) using https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.to_parquet.html

Something like:

```
import io

buffer = io.BytesIO()
df.to_parquet(buffer)
local_array = buffer. getvalue()
```

Reply:
Instead of parquet, went with feather which is supposedly cheaper to write